INT-4214: Migrate to `InvocableHandlerMethod`

artembilan opened the pull request:
JIRA: https://jira.spring.io/browse/INT-4186

* Rework `MessagingMethodInvokerHelper` to delegate to the `InvocableHandlerMethod` with its `HandlerMethodArgumentResolver` infrastructure instead of SpEL
* Introduce several `HandlerMethodArgumentResolver` to address some SI use-cases like `@Payloads`, `@Payload(expression = "")` and `Collection` as argument
* Initialize `DefaultMessageHandlerMethodFactory` in the `MessagingMethodInvokerHelper.start()`.
With that I observed several `Lifecycle` problem when we don't have proper delegate from the top.
* Fix `AbstractCorrelatingMessageHandler` and similar to delegate `Lifecycle` properly
* Fix `ReactiveConsumer` to delegate `Lifecycle` to the `MessageHandler`
* Fix `MutableMessage` do not `generateId()` and set `timestamp` headers if we already have them in the provided headers
* With all that `Lifecycle` many tests must be fixed to call `start()`

artembilan committed:
INT-4186: Migrate to `InvocableHandlerMethod`

JIRA: https://jira.spring.io/browse/INT-4186

* Rework `MessagingMethodInvokerHelper` to delegate to the `InvocableHandlerMethod` with its `HandlerMethodArgumentResolver` infrastructure instead of SpEL
* Introduce several `HandlerMethodArgumentResolver` to address some SI use-cases like `@Payloads`, `@Payload(expression = "")` and `Collection` as argument
* Initialize `DefaultMessageHandlerMethodFactory` in the `MessagingMethodInvokerHelper.start()`.
With that I observed several `Lifecycle` problem when we don't have proper delegate from the top.
* Fix `AbstractCorrelatingMessageHandler` and similar to delegate `Lifecycle` properly
* Fix `ReactiveConsumer` to delegate `Lifecycle` to the `MessageHandler`
* Fix `MutableMessage` do not `generateId()` and set `timestamp` headers if we already have them in the provided headers
* With all that `Lifecycle` many tests must be fixed to call `start()`

artembilan commented:
@garyrussell ,
Take a look, please.
I have spent on this already so much time and still don't see the finish.

* Need to address `Iterator` and `Map` conversion cases.
* Still need to revise the new logic for method analyzing. I feel like we don't need some of the branches.
* I can move init logic outside of `start()` (lazy init in the `process()`) to make test support easier, but i still think that those `Lifecycle`  delegation issues are still valid.
* Would be interested to hear from you if you would like to see some performance comparison test, but unfortunately I dropped SpEL altogether. Hope we still can address conversion concerns via `HandlerMethodArgumentResolver`s 😄

garyrussell commented:
Thanks; I'll take a look tomorrow; my intention was always that we keep SpEL as a fallback option, just in case there is some corner case in user-land that we didn't think of; I still think we will need that; perhaps MIMH2 ? (if it's too complex to keep both).

In any case, I think we can do stand-alone performance analysis.

artembilan commented:
See https://jira.spring.io/browse/SPR-15139 for more info.

artembilan commented:
Since Juergen fixed already https://jira.spring.io/browse/SPR-15139, I think we can go ahead with the solution like falback to SpEL, if `InvocableHandlerMethod` has failed. Just because there is now `MethodArgumentResolutionException` thrown from the `invoke()` when we can't resolve argument for the provided data and method signature.
Any other thoughts?

garyrussell commented:
Sounds good; I haven't had a chance to look at your PR yet.

artembilan committed:
Add SpEL fallback variant to the MessagingMethodInvokerHelper

artembilan commented:
This isn't ready to merge yet. 
Just to share a progress.
Also pay attention to the: https://jira.spring.io/browse/SPR-15151

artembilan committed:
Add `MessagingMethodInvokerHelper.setUseSpelInvoker(boolean)`

Add `MethodInvokingMessageProcessorTests.testPerformanceSpelVersusInvocable()`

artembilan commented:
Provide some further progress after fix https://jira.spring.io/browse/SPR-15151.

There is now performance test-case thanks to new `useSpelInvoker` option.
The result looks like:
```
StopWatch 'SpEL vs Invocable Performance': running time (millis) = 1238
-----------------------------------------
ms     %     Task name
-----------------------------------------
01174  095%  SpEL
00064  005%  Invocable
```

Looks unbelievable 😄 !

So, I might have missed something...

garyrussell commented:
Interesting - would be cool to compare with compiled SpEL too 😄

artembilan commented:
Looks like this:
```
-----------------------------------------
ms     %     Task name
-----------------------------------------
01217  065%  SpEL
00122  007%  Invocable
00535  029%  Compiled SpEL
```

I think we still good to go ahead with all that refactoring hell.
It is worth that 😄 !

If I don't miss something yet...

artembilan committed:
Add Compiled SpEL comparison

garyrussell commented:
Cool - I'll start reviewing after the SCSt calls.

garyrussell commented:
It's not yet clear to me why you need all that lifecycle stuff.

>Initialize DefaultMessageHandlerMethodFactory in the MessagingMethodInvokerHelper.start().

I see that was removed in a later commit.

artembilan commented:
Right, I moved it to the lazy-init part. See my comment https://github.com/spring-projects/spring-integration/pull/2029#issuecomment-272297813.
But we still have to delegate lifecycle control to the target. No?

I maybe would agree that it isn't related to the method invocation. Especially if we use the same `target` for different method invocations. So, `target` lifecycle control should be external.
But! We implemented that part because of https://jira.spring.io/browse/INT-3571.

So, to simplify the review I really can revert that " lifecycle stuff".
But we definitely should revise it afterwards...

artembilan committed:
* Add `MapArgumentResolver` to cover `Properties` case
* add `MessagingMethodInvokerHelper.HandlerMethod.spelOnly` state, when we definitely can perform ony SpEL for provided arguments, e.g. `@Header` with expression
* Catch `IllegalStateException` with the `"argument type mismatch"` message to fallback to SpEL invocation
* Add `Iterator` support for the `CollectionArgumentResolver`
* Add `integrationConversionService` bean registration into the `TestUtils.createTestApplicationContext()`
* Tweak `.travis.yml` to try to download latest JDK, the current `1.8.0_31` is pretty old already and has some bugs
* Adjust some failing tests to use `TestUtils.createTestApplicationContext()` to rely on newly added `integrationConversionService` bean

artembilan committed:
Fix failed tests: `ctx.refresh()`

artembilan committed:
Polishing current year in Copyright

artembilan commented:
The work is done.
Ready for review and merging.

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/handler/support/CollectionArgumentResolver.java:
```diff
@@ -0,0 +1,91 @@
+import org.springframework.util.Assert;
+
+/**
+ * @author Artem Bilan
```
Javadocs

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/handler/support/CollectionArgumentResolver.java:
```diff
@@ -0,0 +1,91 @@
+
+		if (this.canProcessMessageList) {
+			Assert.state(value instanceof Collection,
+					"This Argument Resolver support only messages with payload as Collection<Message<?>>");
```
`s/support only/only supports/`

`s/with payload as/with a payload of/`

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/handler/support/MapArgumentResolver.java:
```diff
@@ -0,0 +1,84 @@
+import org.springframework.messaging.handler.invocation.HandlerMethodArgumentResolver;
+
+/**
+ * @author Artem Bilan
```
Javadoc

garyrussell reviewed (changes requested):
That's all for now - will do a more thorough review tomorrow.

garyrussell commented:
Excellent work!!

One thing that's not clear is how the user practically can call `MIMP.setUseSpelInvoker(true)`.

I am not sure we want to go to the trouble of adding an attribute to every annotation, but on the other hand, setting global option would not be good because it might be just one corner case where they need to force SpEL.

Maybe some kind of registry.

But we can leave this for another PR. Will merge after javadocs are addressed.

artembilan commented:
Yes, I thought about that, Gary, as well.
And my only conclusion to have something like global registry which accepts `MessageHandler` ids to switch to the `useSpelInvoker` in the `afterSingletonsInstantiated()`.
Anyway I think there is still needed some work to pull that `MIMP.setUseSpelInvoker(true)` to the `MH` for proper delegation.
Would be glad to come back to this subject in the separate PR.

Finishing with JavaDocs now...

Meanwhile, please, share your thoughts about `Lifecycle` propagation concern: https://github.com/spring-projects/spring-integration/pull/2029#issuecomment-273321135

artembilan committed:
* JavaDocs for `CollectionArgumentResolver` and `MapArgumentResolver`
* Propagate `ConversionService` from the `MessagingMethodInvokerHelper` to the `MapArgumentResolver.java`

artembilan committed:
Fix `MessagingMethodInvokerHelper` to propagate `BeanFactory` into `MapArgumentResolver` as well
The `MapArgumentResolver` now `extends AbstractExpressionEvaluator`, too
Fix JavaDoc typo

garyrussell commented:
Merged as b3db6a97a25f397bc25c4868422c902035b540fb
